### ISSUE TITLE
Add a `booleanish` option

### ISSUE THREAD

**Issue by jamestalmage** [sindresorhus/meow#44]
This came up while implementing the `--debug` flag for AVA.

Basically, there are two ways to use the debug flag:

```sh
# default port (5858)
ava --debug test.js

# specific port
ava --debug=1234 test.js

# not allowed - this won't work
ava --debug 1234 test.js
```

The reason you can't use the last one is because it's ambiguous whether you meant "run two test files with the default port" or "run one file with port 1234".

When you specify the flag as boolean:

```js
meow(help, {
  boolean: [
    'debug'
  ]
});
```

You get `true` or `false` regardless. However you **must** specify it as a boolean, or this:

```sh
ava --debug test.js
```

To be interpreted as: 

```js
{
  flags: {
    debug: 'test.js'
  }
}
```

My solution was to set `debug` up as a `boolean`, then if I got `true`, reparse argv to find the port

I think it might be interesting to allow this:

```js
meow(helper, {
  booleanish: {
    debug: {
      true: 5858,
      false: false,
      parse: function (value) {
        value = parseInt(value, 10);

        if (isNaN(value)) {
          throw new Error('must be a number');
        }

        return value;
      }
    }
  }
});

The basic usage of the `booleanish` option is:

- `true`: What gets returned if it is used as a boolean flag `--debug`
- `false`: What gets returned if no flag appears (or `--no-debug` is used).
- `parse`: If they use the `--debug=XXX`  - this function is used to parse the `XXX` part. It can throw an Error for invalid values, and `meow` should print the Error message, and exit.

This would greatly ease the creation of dual purpose flags that can be used either as simple booleans, or be supplied with a config value.

**Comment by sindresorhus** [comment]
I like the idea in general and I've had the same problem previously too, but not entirely sold on this exact implementation. Why not just let the user specify an option as both `string` and `boolean` in the minimist options and handle it accordingly? Or maybe introduce a new type in the minimist options called `any` (I prefer it over `booleanish`):

```js
const cli = meow('help', {
  any: ['debug']
});
```

`parse` seems more general and could be added independently of this change, as it could be useful for the `string` type too.

> false: What gets returned if no flag appears (or --no-debug is used).

No flag and `--no-debug` are different. No flag returns `undefined` while `--no-debug` returns `false`. This distinction is usually not important, but can be.

-

// @kevva @SamVerschueren

**Comment by Exagram** [comment]
I agree with this direction. General unix programs have inputs such as "mysql -u Eminem -h ****.org". With meow (which Yeoman uses) this is impossible! Instead you are forced to use equals: mysql -username=Eminem -host=****.org. You can't even use the short-hand aliases.

**Comment by sindresorhus** [comment]
@Exagram I don't see how your comment is relevant to this issue? You can already create short-hand flags using the [minimist `alias` option](https://github.com/sindresorhus/meow#minimistoptions).

**Comment by Mouvedia** [comment]
> Why not just let the user specify an option as both string and boolean in the minimist options and handle it accordingly?

Is this supported?

**Comment by ittaiz** [comment]
👋 just stumbled on this issue after I had the need to have a --branch argument which can be a boolean and mean use a generated branch name or --branch=foo and mean that the branch name is foo

**Comment by sindresorhus** [comment]
This will have to be supported in `yargs-parser` first.

Slightly relevant issue: https://github.com/yargs/yargs/issues/1599
